packages/cli/src/modules/workflow-index/workflow-index.service.ts

```diff
@@ -0,0 +1,165 @@
+import { Logger } from '@n8n/backend-common';
+import { WorkflowDependency, WorkflowDependencyRepository, WorkflowRepository } from '@n8n/db';
+import { Service } from '@n8n/di';
+import { IWorkflowBase } from 'n8n-workflow';
+
+import { EventService } from '@/events/event.service';
+
+@Service()
+export class WorkflowIndexService {
+	constructor(
+		private readonly eventService: EventService,
+		private readonly dependencyRepository: WorkflowDependencyRepository,
+		private readonly workflowRepository: WorkflowRepository,
+		private readonly logger: Logger,
+	) {}
+
+	init() {
+		this.eventService.on('server-started', async () => {
+			this.logger.info('Building workflow dependency index...');
+			// eslint-disable-next-line @typescript-eslint/return-await
+			return this.buildIndex();
+		});
+		this.eventService.on('workflow-created', async ({ workflow }) => {
+			await this.updateIndexFor(workflow);
+		});
+		this.eventService.on('workflow-saved', async ({ workflow }) => {
+			await this.updateIndexFor(workflow);
+		});
+		this.eventService.on('workflow-deleted', async ({ workflowId }) => {
+			// Updating with an empty list effectively removes the existing dependencies.
+			await this.dependencyRepository.updateDependenciesForWorkflow(workflowId, []);
+		});
+		this.eventService.on('workflow-archived', async ({ workflowId }) => {
+			// Updating with an empty list effectively removes the existing dependencies.
+			await this.dependencyRepository.updateDependenciesForWorkflow(workflowId, []);
+		});
+		this.eventService.on('workflow-unarchived', async ({ workflowId }) => {
+			await this.workflowRepository.findOneBy({ id: workflowId }).then(async (workflow) => {
+				if (workflow) {
+					await this.updateIndexFor(workflow);
+				}
+			});
+		});
+	}
+
+	private async buildIndex() {
+		// Get all the workflows.
+		const workflows = await this.workflowRepository.find();
```

Comment:
To reduce data, since we only need nodes, the ID and counter, we could select only those and not e.g. connections